README
See the file BUILDING for building instructions.  The setup program makes 
configuration files based on machine-specific details in the "machines" 
directory.

Upon unpacking the source, you will find the following directories:

autodoc         Autodoc files for programs using PIP package for input
clip		Source for Command Line Image Processing
com		Command files, mostly used by copytomocoms for tomography
ctfplotter      Source for ctfplotter program
dist		Files for installation and distribution
Etomo           eTomo package tree
Etomo/doc       Program documentation files
Etomo/scripts   Has an eTomo testing script
Etomo/src       Top of source tree
Etomo/src/etomo All java source is under here
  .../comscript Command script classes
  .../logic     
  .../process   Process-running classes
  .../storage   File-related classes
  .../type      Classes to handle specific data items
  .../ui/swing  Dialog classes
  .../uitest    Classes for the UI testing
  .../util      Utility functions
Etomo/src/images  Has icons
flib		Fortran program source tree
flib/beadtrack	Source for bead tracker program
flib/blend	Source for montage blending program and montage reducer
flib/distort	Source for distortion-finding program
flib/graphics	Source for several programs using line graphics
flib/image	Source for many programs that do not use models or graphics
flib/man	Master directory for man pages of Fortran programs
flib/model	Source for many programs that use IMOD models
flib/ndasda	Source for nda, sda, and mtk spatial analysis programs
flib/subrs	Top of source tree for Fortran libraries
  .../blas      Source for blas routines needed by lapack
  .../compat    Source for compatibility routines where needed
  .../dtrigs	Source for cosd, sind, etc. functions for Linux (libdtrigs)
  .../graphics	Source for graphics subroutines (libdnmncar)
  .../hvem	Source for miscellaneous subroutines developed here
  .../ifftsub	Source for FFT subroutines  (libifft - no longer used)
  .../imsubs	Source for basic image I/O routines  (libim)
  .../imsubs2	Source for other image manipulation routines  (libim)
  .../lapack    Source for subset of lapack needed by ctfplotter
  .../linetrack	Source for line tracker subroutines
  .../model	Source for model-related subroutines  (libwmod)	
  .../piecesubs	Source for routines that work with piece coordinates  (libhvem)
  .../randm	Source for Linux replacement for RAN function  (librandm)
  .../smallmodel  Derived source for library that has a small model (libsmlmod)
  .../statsubs	Source for statistical subroutines
  .../xfsubs	Source for routines that work with transformations
flib/tilt	Source for back-projection program
flib/tilt/fbplib  Source for library of routines for fast back-projection
flib/tiltalign	Source for tilt alignmnet program
html		All html documents and files that they link to
html/3dmodHelp  Help files for 3dmod
html/3dmodimages Images for introduction to 3dmod
html/ctfHelp    Help files for ctfplotter
html/etomoImages Images for etomo tutorial
html/joinImages  Images for join tutorial
html/libdoc     Framework files for library documentation
html/midasHelp  Help files for midas
3dmod		Source for 3dmod program
imodutil	Source for utility programs for working with IMOD models
include		Include files for IMOD programs
libcfft         Source for C version of FFT library
libcfshr        Source for library of utilities shared with Fortran
libdiaqt        Source for Qt dialog and utility library
libiimod        Source for IMOD image library
libimod         Source for IMOD model library
libmesh         Source for meshing library
libwarp         Source for warping library, includes natural neighbor 
                       interpolation and Delaunay triangulation code
machines	Machine-specific files included by setup script
manpages	Master source for man pages of C programs, in nroff format
midas		Source for midas program
mrc		Source for various programs manipulating MRC image files
plugs		Source tree for imod plugins
plugs/analysistools  Source for Andrew Noske's analysis tools plugin
plugs/beadfix	   Source for old version of bead fixer plugin used to test build
plugs/beadhelper   Source for Noske's bead helper plugin
plugs/drawingtools Source for Noske's drawing tools plugin
plugs/interpolator Source for Noske's interpolator plugin
plugs/namewizard   Source for Noske's namewizard plugin
plugs/noskecommon  Source for common files for Noske's plugins
plugs/stereology   Source for Noske's stereology plugin
processchunks   Source for processchunks program
pysrc           Python scripts
qtassist        Source for imodqtassist program
raptor          Source for RAPTOR
scripts		Shell scripts
sendevent       Source for imodsendevent program
sourcedoc       Source for sourcedoc program (Qt-based)
sysdep          System dependent files needed for builds
sysdep/mac      Files for OS X build - not currently needed
sysdep/win      Files for Windows build - Tiff-related files and manifests


After doing a "make" and default "make install", you will have:

buildlib	Newly built but uninstalled libraries, needed to make programs
man		Top directory for man pages (contains a cat1 and cat5)
html/man	Html versions of man pages
html/libhelp    Processed library documentation files
html/plughelp   Copies of plugin help files, needed for inclusion in IMOD.qch
bin		Installation directory for executable programs and scripts
lib		Installation directory for libraries (old 32-bit on SGI)
lib/imodplug	Installation directory for imod plugins
pylib           Installation directory for Python modules

Other files in top directory:
BUILDING        Details for building IMOD on all platforms
History         Has the complete history of changes
installqtlib    Does post-install copying of libraries and startup files
Makefile        Master Makefile for build
packMacApps     For a Mac install, puts Qt programs into .app directories
setup           Starting script for configuring for build
setup2          Real configure script; setup runs dos2unix on it in Cygwin
vcimod.dsw      Workspace for midas, 3dmod, imodsendevent in Visual C++


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3/29/14: In eTomo - PEET, fixed the loading of a non-array selectClassID
value, and fixed more problems with unnecessary errors generated because of
missing parameters (In 4.7).(Bug# 1789)

3/27/14: Fixed a communication problem between eTomo and 3dmod and rolled back
previous incorrect fixes from 2/14/14 and 3/14/14 (In 4.7).(Bug# 1792)

3/27/14: Fixed HDF writing, optimized chunk cache size for volume writing,
fixed pyramid creation, made 3dmod recognize tiling in HDF file.

3/24/14: IMOD_4-8-2
3/23/14: Eliminated an input routine that reads line-by-line, which was giving
terrible 3dmod performance on Mac laptops loading over a network.

3/22/14: Programs can now read and write HDF files interchangeably with MRC
files. Write performance is currently atrocious.  Metadata can be stored as
well; Newstack will move metadata between an HDF file and an .mdoc file
associated with another file type, or even between two mdoc files.  Supported
HDF files can consist of stacks of 2D images of the same size, or files with
multiple 3D volumes, which 3dmod will open like multiple files.  Makepyramid
can make a multi-volume HDF file that 3dmod will recognize as an image
pyramid.
3/22/14: Fixed newstack when -fromone used with default section list (In 4.7).

3/14/14: Fixed a Windows problem with opening 3dmod from eTomo where
imodsendevent was called incorrectly (In 4.7).(Bug# 1786)

3/14/14: Made the close button on the detached toolbar of linked slicers close
all the linked slicers (In 4.7).

3/13/14: Fixed a problem in eTomo - PEET where tilt angles where not being
loaded from a renamed tilt log file (In 4.7).(Bug# 1782)

3/8/14: Fixed deferring of initial messages to 3dmod on Mac (In 4.7).

3/8/14: Fixed lack of title bar and initial placement of toolbar for linked 
slicers (In 4.7).

3/7/14: Updated the Join tutorial (In 4.7).(Bug# 1785)

3/5/14: IMOD_4-7-1
3/5/14: In eTomo - PEET, made sure that empty attributes will not generate
Matlab syntax errors.(Bug# 1784)

2/28/14: IMOD_4-8-0 and branch to IMOD_4-7-0
2/28/14: Reading the bidirectional tilt value from the header file.(Bug# 1781)

2/26/14: Add options to 3dmod for setting initial center, angles, and zoom.

2/22/14: Fix exit status, button width, splash screen for Ctfplotter on Mac.

2/21/14: Added three more colors for arrows drawn with Grab with Note.

2/21/14: Fixed changing transparency during model view movies.

2/15/14: IMOD_4-6-41
2/14/14: In eTomo, fixed a problem with slowness when opening 3dmod dialogs.
(Bug# 1778)

2/14/14: In eTomo - PEET fixed problems with defaults and the handling of
missing properties.(Bug# 1738)

2/13/14: Switched default antialiasing filter for Newstack to Lanczos 3.

2/9/14: Updated the images in the eTomo tutorial.  Also exposed the
autofidseed append parameter and initialized the seed model contours in 3dmod.
(Bug# 1777)

2/9/14: In eTomo - Setup Reconstruction, allowing .mrc image files to be
selected, and renaming them to .st.(Bug# 1696)

2/7/14: Fixed problems with installer on Python 2.5 and Python 3.

2/6/14: In eTomo - PEET, accepting integer vectors in selectClassID.
(Bug# 1775)

2/5/14: Makecomfile adds option to append to seed model for autofidseedb.com.

2/3/14: Changed Bead Fixer so value-based controls show or hide appropriately
when model is changed.

2/1/14: Made installer able to take a .csh file, so it can be used to upgrade
on Windows without Cygwin.

1/31/14: Fixed installer test for 32/64 bit mismatch and added test and
message about libjpeg 6.2 on Ubuntu.

1/30/14: IMOD_4-6-40
1/30/14: Switched to outputting signed bytes by default.  BYTE FILES WILL NO
LONGER BE READABLE BY VERSIONS BEFORE IMOD 4.3.

1/30/14: Fixed execute permission on Autofidseed and added option for passing
options to Imodfindbeads.

1/30/14: In eTomo fixed bugs in the Directive Editor.(Bug# 1721)

1/30/14: In eTomo - Directive Editor, fixed bugs which where causing unchanged
directives to be display in the default mode.(Bug# 1772)

1/30/14: In eTomo - Directive Editor, getting CTF defaults from origcoms.
(Bug# 1715)

1/29/14: IMOD_4-6-39
1/29/14: Changed Newstack floating to truncate images with extreme ranges
instead of scaling them to a different mean/SD.

1/29/14: In eTomo - Coarse Alignment, fixed a problem with graphing edge
errors.(Bug# 1774)

1/29/14: In eTomo improving error handling for an empty fiducial diameter
field.(Bug# 1671)

1/29/14: Fixed problems with autofidseed in eTomo - Fiducial Model Dialog.
(Bug# 1743)

1/29/14: Exposed flgNoReferenceRefinement in eTomo - PEET.(Bug# 1738)

1/28/14: Restored fortran graphics program wrappers to Mac build, lost 1/13.

1/20/14: IMOD_4-6-38
1/20/14: Fixed keypad hot keys in Model View on Mac.

1/18/14: IMOD_4-6-37

1/18/14: Further polishing of Grab with Note, which can now be used to draw
multiple arrows in Zap or Slicer windows.

1/16/14: IMOD_4-6-36
1/16/14: Jane Ding contributed Grab with Note plugin, which takes Slicer or Zap
snapshots with notes added below the image.

1/16/14: Fixed creation of patch tracking com file if mag change searched in
coarse align.

1/16/14: Change eTomo to make it independent of Cygwin.  Fixed a problem with
killing processes on Windows.(Bug# 1716)

1/14/14: Fixed bug in clip multiplying while unpacking 4-bit data, added
rounding in clip multiply or divide when outout is to an integer.

1/12/14: Made Imodchopconts able to chop up or sort out contours based on
fine-grained color.

1/11/14: Fixed 3dmod crash when the I button is pressed in Zap and no image
loaded, present since 4.6.29.

1/10/14: Made 3dmod able to read single pixels from non-MRC-like files so
Pixel View and Graph windows can show file values.
 
1/8/14: Made all C/C++ programs except Raptor able to read TIFF and MRC-like
file formats, made it possible to direct output to a format specified by
environment variable. TIFF stacks are now written with min and max of the file

imod/imod.h
/*  imod.h - public header file for 3dmod
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2007 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 *  Log at end of file
 */
#ifndef IMOD_H
#define IMOD_H

class QKeyEvent;

#ifndef IMODP_H
#include "imodP.h"
#include "imod_utilities.h"
#endif

/* Define macro for export of functions under Windows */
#ifndef DLL_EX_IM
#ifdef _WIN32
#define DLL_EX_IM _declspec(dllexport)
#else
#define DLL_EX_IM
#endif
#endif

/*********************** Graphics functions. *********************************/
/* DOC_SECTION DEFINES */
/* DOC_CODE Draw flags */
/* 
 * Defines for imodDraw() and ivwDraw().  IMOD_DRAW_ACTIVE and IMOD_DRAW_TOP
 * are flags added by the control module when calling the active or top window
 * and would not be included in calls to the draw routines.
 */

/* image data or color map has changed; windows need to clear caches and draw
   all images */
#define IMOD_DRAW_IMAGE       (1) 

/* The current point x, y, z position has been changed.  This causes an update
   of the info window x/y/z, image scale dialog if multifile Z loaded, the 
   image processing window, and the isosurface window.  Model view will be 
   drawn if the image display is on, unless the IMOD_DRAW_SKIPMODV flag is 
   also set. */
#define IMOD_DRAW_XYZ      (1<<1) 

/* Model data or current model object/contour/point has been changed. 
   Prior to XYZ update, this causes update to info window obj/cont/pt, and
   to object edit, cont/surf/pt, contour move and fine grain dialogs and
   update of plugins with IMOD_REASON_MODUPDATE.  Model view will be drawn
   unless the IMOD_DRAW_SKIPMODV flag is also set. */
#define IMOD_DRAW_MOD      (1<<2)

/* A slicer location has changed; Zap and XYZ windows will draw slicer lines */
#define IMOD_DRAW_SLICE    (1<<3) 

/* Skip drawing model view.  Use this to cause a draw of image windows when
 a model view draw is already inevitable */
#define IMOD_DRAW_SKIPMODV (1<<5) 

/* Color index mode color map has changed.  Do not combine with other flags. */
#define IMOD_DRAW_COLORMAP (1<<11)

/* Do not resync image to model point.  This keeps Zap from recentering. */
#define IMOD_DRAW_NOSYNC  (1<<12)

/* Recalculate cursor position: adds the IMOD_DRAW_MOD flag, and if there is
   a current model point defined, sets current point x,y,z to it and adds the 
   IMOD_DRAW_XYZ flag */
#define IMOD_DRAW_RETHINK (1<<13)
#define IMOD_DRAW_ACTIVE  (1<<14) /* current window is active.            */
#define IMOD_DRAW_TOP     (1<<15) /* current window has highest priority. */

#define IMOD_DRAW_ALL (IMOD_DRAW_IMAGE | IMOD_DRAW_XYZ | IMOD_DRAW_MOD)
/* END_CODE */
/* END_SECTION */

#include "imodview.h"

#define IMOD_MM_TOGGLE 0

extern "C" {

/*!
 * Draw [model] using a 2D line renderer implemented with OpenGL functions.
 * This is the function used to draw the slicer window.  You will need to set
 * up the view matrix yourself, including Z limits in the projection to avoid
 * getting the whole model drawn on a slice.  Set [drawCurrent] nonzero to
 * have current image or model point and current contour end symbols drawn too.
 * Provide the z-scaling used to render the image in [zscale].
 */
void DLL_EX_IM imodDrawModel(ImodView *inImodView, Imod *inModel, 
                             int drawCurrent, float zscale);
/* 3dmod plugin */
/************************************* defines *******************************/
/* DOC_SECTION DEFINES */
/* DOC_CODE Plugin flags */
/* The IMOD_PLUG_ flags are ones that a plugin sends back in response to
 * the ImodPlugInfo call.  The IMOD_REASON_ values are sent to the plugin in 
 * the imodPlugExecute call 
 */
#define IMOD_PLUG_MENU     1  /* Add to special plugin menu. */
#define IMOD_PLUG_TOOL     2  /* Add to toolbar (not implemented)     */
#define IMOD_PLUG_PROC     4  /* Add to image proc window. (not implemented) */
#define IMOD_PLUG_VIEW     8  /* Add to model view window. (not implemented) */
#define IMOD_PLUG_KEYS    16  /* Handle key events. */
#define IMOD_PLUG_FILE    32  /* Allow other image files to be loaded (?). */
#define IMOD_PLUG_MESSAGE 64  /* Execute messages */
#define IMOD_PLUG_MOUSE  128  /* Handle mouse events */
#define IMOD_PLUG_EVENT  256  /* Handle other events (wheel, enter, leave) */

#define IMOD_REASON_EXECUTE 1 /* Execute plugin after selection from menu  */
#define IMOD_REASON_STARTUP 3 /* 3dmod has started. Initialize plugin. */
#define IMOD_REASON_MODUPDATE 4 /* Update the plugin for model changes */
#define IMOD_REASON_NEWMODEL  5 /* Update the plugin for new model loaded */
/* END_CODE */


/**************************** Application Data *******************************/

/* DOC_CODE Color defines */
/* color defines. */
#define COLOR_BACKGROUND 1  /* background color */
#define COLOR_FOREGROUND 2  /* foreground color */
#define COLOR_SELECT     3  /* selection color  */
#define COLOR_SHADOW     4  /* shadowed selection color */
#define COLOR_END        5  /* endpoint color   */
#define COLOR_BEGIN      6  /* begin point color */
#define COLOR_POINT      7  /* point color */
#define COLOR_GHOST      8  /* ghost color, darker version of object. */
#define COLOR_MIN        9  /* The index of the minimum image value. */
#define COLOR_MAX       10  /* The index of the maximum image value. */
/* END_CODE */
/* END_SECTION */

/*!
 * Gets the color depth; 24 on current systems with no color index mode *
 */
int DLL_EX_IM imodDepth(void);

/*!
 * Returns the pixel value of the given COLOR_ value in [inColor], used to 
 * call b3dColorIndex
 */
int DLL_EX_IM imodColorValue(int inColor);


/************************ utility functions. *********************************/

/*! Print text to the 3dmod information window.
 * The usage of this function is similar to the usage of the
 * stdio function printf.
 */
void DLL_EX_IM wprint(const char *fmt, ...);

/*! Prints an error message with printf-type arguments
   This function will output a message box if [out] is NULL or under Windows;
   otherwise it will print to [out], which should be {stdout} or {stderr} */
void DLL_EX_IM imodError(FILE *out, const char *format, ...);

/*! Prints a message to standard error with printf-type arguments
  This function will flush stderr under Windows */
void DLL_EX_IM imodPrintStderr(const char *format, ...);

/*! Puts [message] to standard error like puts, but flushes under Windows */
void DLL_EX_IM imodPuts(const char *message);

/*! Prints [message] to standard out, or to a message box under Windows */
void DLL_EX_IM imodPrintInfo(const char *message);


/*! Passes [even] to the general key handler to execute 3dmod hot keys */
void DLL_EX_IM imodDefaultKeys(QKeyEvent *event, ImodView *vw);

/*! Shows a help page with filename [page] in Qt Assistant; provide a full
 * path if the path is not relative to IMOD_DIR/html/3dmodHelp.
 * Returns 1 for error, 0 for success 
 */
int DLL_EX_IM imodShowHelpPage(const char *page);


/*****************************************************************************/
}
#endif /* IMOD_H */

/*
    $Log$
    Revision 3.19  2008/05/28 14:57:28  mast
    Added z scale to draw model function

    Revision 3.18  2008/05/22 20:58:33  mast
    Documentation of IMOD_DRAW flags

    Revision 3.17  2008/02/06 16:14:09  sueh
    bug# 1065 Changed wprint(char *) to wprint(const char*).

    Revision 3.16  2008/01/17 22:33:54  mast
    Added reason define for plugin update call

    Revision 3.15  2008/01/14 19:46:55  mast
    moved toggle flag to public file

    Revision 3.14  2007/12/04 22:02:41  mast
    Changes for documentation

    Revision 3.13  2007/11/30 06:51:50  mast
    Changes for linking slicer to model view

    Revision 3.12  2007/06/04 15:02:33  mast
    Added argument to draw model function

    Revision 3.11  2006/02/13 05:13:36  mast
    Define mouse flag for plugins

    Revision 3.10  2004/12/24 02:19:11  mast
    Removed argument from help page call

    Revision 3.9  2004/11/22 00:22:48  mast
    Added definition for showing help page

    Revision 3.8  2004/09/24 18:10:04  mast
    Added new definition for plugins that can execute messages

    Revision 3.7  2004/05/31 23:10:56  mast
    Added macros for exporting/importing under Windows

    Revision 3.6  2003/10/01 05:04:44  mast
    Changes for creation of imodview.h

    Revision 3.5  2003/09/16 02:49:06  mast
    Changed declarations ofr functions that return image line pointers

    Revision 3.4  2003/06/27 19:24:45  mast
    Add function to get extra object

    Revision 3.3  2003/02/27 18:13:59  mast
    Add ImodView argument to imodDrawModel so that times can be detected

    Revision 3.2  2003/02/10 20:41:55  mast
    Merge Qt source

    Revision 3.1.2.4  2003/01/29 01:43:00  mast
    add colormap draw flag

    Revision 3.1.2.3  2003/01/27 00:30:07  mast
    Pure Qt version and general cleanup

    Revision 3.1.2.2  2003/01/13 01:15:42  mast
    changes for Qt version of info window

    Revision 3.1.2.1  2003/01/02 15:38:16  mast
    remove declarations for control.c functions

    Revision 3.1  2002/12/01 15:34:41  mast
    Changes to get clean compilation with g++

*/
